host/client_integration_test.go

```diff
@@ -599,6 +599,113 @@ func (s *clientIntegrationSuite) Test_ActivityTimeouts() {
 	//s.printHistory(id, workflowRun.GetRunID())
 }
 
+func (s *clientIntegrationSuite) Test_UnhandledCommandAndNewTask() {
+	/*
+		Event sequence:
+		1st WorkflowTask runs a local activity.
+		While local activity is running, a signal is received by server.
+		After signal is received, local activity completed, and workflow drains signal chan (no signal yet) and complete workflow.
+		Server failed the complete request because there is unhandled signal.
+		Server rescheduled a new workflow task.
+		Workflow runs the local activity again and drain the signal chan (with one signal) and complete workflow.
+		Server complete workflow as requested.
+	*/
+
+	sigReadyToSendChan := make(chan struct{}, 1)
+	sigSendDoneChan := make(chan struct{})
+	localActivityFn := func(ctx context.Context) error {
+		// to unblock signal sending, so signal is send after first workflow task started.
+		select {
+		case sigReadyToSendChan <- struct{}{}:
+		default:
+		}
+
+		// this will block workflow task and cause the signal to become buffered event
+		select {
+		case <-sigSendDoneChan:
+		case <-ctx.Done():
+		}
+
+		return nil
+	}
+
+	var err1 error
+	var receivedSig string
+	workflowFn := func(ctx workflow.Context) error {
+		ctx1 := workflow.WithLocalActivityOptions(ctx, workflow.LocalActivityOptions{
+			StartToCloseTimeout: 10 * time.Second,
+			RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 1},
+		})
+		f1 := workflow.ExecuteLocalActivity(ctx1, localActivityFn)
+		err1 = f1.Get(ctx1, nil)
+		if err1 != nil {
+			return err1
+		}
+
+		sigCh := workflow.GetSignalChannel(ctx, "signal-name")
+
+		for {
+			var sigVal string
+			ok := sigCh.ReceiveAsync(&sigVal)
+			if !ok {
+				break
+			}
+			receivedSig = sigVal
+		}
+
+		return nil
+	}
+
+	s.worker.RegisterWorkflow(workflowFn)
+
+	id := "integration-test-unhandled-command-new-task"
+	workflowOptions := sdkclient.StartWorkflowOptions{
+		ID:                  id,
+		TaskQueue:           s.taskQueue,
+		WorkflowRunTimeout:  10 * time.Second,
+		WorkflowTaskTimeout: 10 * time.Second,
```

Comment:
nit: is my understanding correct that taskTimeout here must equal to runTimeout? Otherwise workflow may still succeed after the workflow task timeout? If so, probably add a comment here?

Reply:
That is correct. The timeout was intentionally set to the same so if workflow task timeout it would fail the workflow. I could add a comment to clarify.

Reply:
But the test also verify the workflow history events.